ext/mbstring/libmbfl/mbfl/mbfilter.c

```diff
@@ -93,2565 +92,1438 @@
 	}
 
-	if (transenc->no_encoding == mbfl_no_encoding_qprint) {
-		pe->encod_filter->status |= MBFL_QPRINT_STS_MIME_HEADER;
-		pe->encod_filter_backup->status |= MBFL_QPRINT_STS_MIME_HEADER;
-	} else {
-		pe->encod_filter->status |= MBFL_BASE64_STS_MIME_HEADER;
-		pe->encod_filter_backup->status |= MBFL_BASE64_STS_MIME_HEADER;
+	int line_length = pe->outdev.pos - pe->linehead + pe->encname_len + pe->firstindent;
+	if (line_length > 60)  {
+		mime_header_break_line(pe);
+	} else if (pe->outdev.pos > 0)  {
+		EMIT_SPACE();
 	}
 
-	return pe;
-}
+	mbfl_memory_device_strncat(&pe->outdev, pe->encname, pe->encname_len);
 
-void
-mime_header_encoder_delete(struct mime_header_encoder_data *pe)
-{
-	if (pe) {
-		mbfl_convert_filter_delete(pe->conv1_filter);
-		mbfl_convert_filter_delete(pe->block_filter);
-		mbfl_convert_filter_delete(pe->conv2_filter);
-		mbfl_convert_filter_delete(pe->conv2_filter_backup);
-		mbfl_convert_filter_delete(pe->encod_filter);
-		mbfl_convert_filter_delete(pe->encod_filter_backup);
-		mbfl_memory_device_clear(&pe->outdev);
-		mbfl_memory_device_clear(&pe->tmpdev);
-		efree((void*)pe);
+	while (p < e) {
+		output_function(*p++, &converted, conv2_filter, encode_filter, pe);
 	}
+
+	(conv2_filter->filter_flush)(conv2_filter);
+	mbfl_convert_filter_devcat(encode_filter, &converted);
+	(encode_filter->filter_flush)(encode_filter);
+
+	mbfl_memory_device_strncat(&pe->outdev, "?=", 2);
+
+	mbfl_memory_device_clear(&converted);
+	mbfl_convert_filter_delete(conv2_filter);
+	mbfl_convert_filter_delete(encode_filter);
 }
 
-mbfl_string *
-mbfl_mime_header_encode(
-    mbfl_string *string,
-    mbfl_string *result,
-    const mbfl_encoding *outcode,
-    const mbfl_encoding *encoding,
-    const char *linefeed,
-    int indent)
+mbfl_string *mbfl_mime_header_encode(mbfl_string *string, mbfl_string *result, const mbfl_encoding *outcode,
+	const mbfl_encoding *transfer_encoding, const char *linefeed, int indent)
 {
-	size_t n;
-	unsigned char *p;
-	struct mime_header_encoder_data *pe;
-
-	mbfl_string_init(result);
-	result->encoding = &mbfl_encoding_ascii;
+	mbfl_string_init_set(result, &mbfl_encoding_ascii);
 
-	pe = mime_header_encoder_new(string->encoding, outcode, encoding);
+	struct mime_header_encoder_data *pe = mime_header_encoder_new(string->encoding, outcode, transfer_encoding);
 	if (pe == NULL) {
 		return NULL;
 	}
 
-	if (linefeed != NULL) {
-		n = 0;
-		while (*linefeed && n < 8) {
-			pe->lwsp[n++] = *linefeed++;
-		}
-		pe->lwsp[n++] = 0x20;
-		pe->lwsp[n] = '\0';
-		pe->lwsplen = n;
+	if (linefeed) {
+		pe->linefeed_len = snprintf(pe->linefeed, sizeof(pe->linefeed), "%.8s ", linefeed);
 	}
 	if (indent > 0 && indent < 74) {
 		pe->firstindent = indent;
 	}
 
-	n = string->len;
-	p = string->val;
-	while (n > 0) {
-		(*pe->conv1_filter->filter_function)(*p++, pe->conv1_filter);
-		n--;
+	size_t len;
+	unsigned int *wc_buffer = convert_string_to_wchar(string, &len), *p = wc_buffer, *e = p + len, *saved = p;
```

Comment:
Very true. I also thought about that, but opted to allocate the buffer since it made the code a lot nicer to read. And although this does add some overhead, on the whole the function is still faster than it used to be.